iob_soc.py

```diff
@@ -40,308 +43,606 @@ class iob_soc(iob_module):
-        uart.instance("UART0"),
-        timer.instance("TIMER0"),
-    ]
-    peripheral_portmap = [  # List of tuples, each tuple corresponds to a port map
-        (
-            {
-                "corename": "UART0",
-                "if_name": "rs232",
-                "port": "txd",
-                "bits": [],
-            },
-            {
-                "corename": "external",
-                "if_name": "uart",
-                "port": "uart_txd_o",
-                "bits": [],
-            },
+            descr="Bus of AXI master interfaces for external memory. One interface for this system and others optionally for peripherals.",
+            if_defined="USE_EXTMEM",
+            ports=[],
+            connect_to=axi
         ),
-        (
-            {
-                "corename": "UART0",
-                "if_name": "rs232",
-                "port": "rxd",
-                "bits": [],
-            },
-            {
-                "corename": "external",
-                "if_name": "uart",
-                "port": "uart_rxd_i",
-                "bits": [],
-            },
+        iob_interface(
+            name="rs232",
+            type="master",
+            port_prefix="",
+            wire_prefix="",
+            descr="iob-soc uart interface",
+            ports=[
+                iob_port(
+                    name="txd",
+                    direction="output",
+                    width=1,
+                    descr="UART transmit data",
+                    connect_to=txd,
+                ),
+                iob_port(
+                    name="rxd",
+                    direction="input",
+                    width=1,
+                    descr="UART receive data",
+                    connect_to=rxd,
+                ),
+                iob_port(
+                    name="rts",
+                    direction="output",
+                    width=1,
+                    descr="UART request to send",
+                    connect_to=rts,
+                ),
+                iob_port(
+                    name="cts",
+                    direction="input",
+                    width=1,
+                    descr="UART clear to send",
+                    connect_to=cts,
+                ),
+            ],
         ),
-        (
-            {
-                "corename": "UART0",
-                "if_name": "rs232",
-                "port": "cts",
-                "bits": [],
+    ]
+
+    #######################################
+    # IOb-SoC modules, wires, and instances
+    #######################################
+
+    #
+    # SYSTEM RESET
+    #
+
+    boot = iob_wire("boot", width=1)
+    cpu_reset = iob_wire("cpu_reset", width=1)
+
+    #
+    # CPU
+    #
+
+    cpu_i_req = iob_wire("cpu_i_req", width=REQ_W)
+    cpu_i_resp = iob_wire("cpu_i_resp", width=RESP_W)
+    cpu_d_req = iob_wire("cpu_d_req", width=REQ_W)
+    cpu_d_resp = iob_wire("cpu_d_resp", width=RESP_W)
+
+    cpu = iob_picorv32(
```

Comment:
As requested, the new constructors of modules no longer create a "verilog module" or a "verilog instance". Now they create an object that represents a mix between the two.

In this line 310, we are creating a Verilog module and also a Verilog instance, since we are giving it Verilog parameters, and connecting its IOs. We will also instantiate it in the iob_soc.v, by [placing it in the `blocks` list.](https://github.com/IObundle/iob-soc/pull/803/files#diff-551ef7ec958375dc7823b60fea7a6b8ff010a80b79c53afdd937d66369e41807R633).

However, in [lines 601-626](https://github.com/IObundle/iob-soc/pull/803/files#diff-551ef7ec958375dc7823b60fea7a6b8ff010a80b79c53afdd937d66369e41807R601-R626), we are just creating Verilog modules, not Verilog instances. These modules are required because they are used elsewhere in the system (when a submodule needs them but does not include them), or they cannot be directly instantiated as Verilog instances (like verilog headers and software sources).